### PR TITLE
feat: change state to userdict

### PR DESCRIPTION
**Pull request by younesStrittmatter**
# Description

Change state into a user dict

- resolves https://github.com/AutoResearch/autora/issues/548

# Type of change
- **feat**: A new feature

- **BREAKING_CHANGE**: Not sure if it breaks existing code, but tests/notebooks in the core worked fine

# Remarks (Optional)
- The old state class is still implemented and State, StandardState are now aliases of the "Dict"-Versions. If we find Problems, we can easily go back to the Dataclass solution
- The advantage of this is, that the fields of the state can now be changed dynamically. For example, one can add fields easily and alter their "delta"-behaviour between cycles

### CODE REVIEW

**Review by musslick — CHANGES_REQUESTED** [review]
This looks great! The only suggestion I have is to add a minimal jupyter notebook under docs/cycle that outlines the new functionality, i.e., how the state can be modified (adding fields, changing delta functions).

**Review by musslick — APPROVED** [review]
Looks great!

File: src/autora/state.py [hunk 1]
```diff
@@ -45,8 +45,95 @@ def __add__(self: C, other: Union[Delta, Mapping]) -> C:
 S = TypeVar("S", bound=DeltaAddable)
 
 
+class StateDict(UserDict):
```
Comment (hollandjg) [inline src/autora/state.py:48]: I think this class would benefit from a bunch of doctests which demo and check it's full functionality! You could start with the doctests in the StandardState as an example, and then add the new functionality too.

**Review by hollandjg — CHANGES_REQUESTED** [review]
I think this needs a bunch of automated tests before it gets merged — it's such a central part of the core, that we really need confidence it's working!

**Pull request merged**